runtime/vm/LayoutFFITypeHelpers.hpp

```diff
@@ -420,11 +458,23 @@ class LayoutFFITypeHelpers
 		 */
 		parseStruct(&structSig, &isAllSP, &isAllDP, first16ByteComposTypes, &curIndex);
 
-		if (isAllSP) {
+		/* As per the z/OS ABI, ALL_SP/ALL_DP are only intended for struct {float, float} and
+		 * struct {double, double} or their variants as the complex type placed on FPRs given
+		 * the struct size is 8 bytes for [FF]/variant or 16 bytes for [DD]/variant; otherwise,
+		 * the non-complex types are categorized into AGGREGATE_OTHER placed on GPRs.
+		 *
+		 * e.g.
+		 * 1) struct {float, union {float, float}} belong to ALL_SP while
+		 *    struct {double, union {double, double}} belongs to ALL_DP.
+		 * 2) struct {float a[2]} ([2:F] in terms of the layout string)
+		 *    and struct {double b[2]} ([2:D] in terms of the layout string)
+		 *    belong to AGGREGATE_OTHER placed on GPRs.
+		 */
+		if (isAllSP && isStructWithFFOrDD(initSig, sizeInByte, J9_FFI_UPCALL_STRUCT_FF_SIZE)) {
```

Comment:
1) as far as i know, literally only [float, float] and [double, double] are considered as complex on zOS. any other (e.g. [float, {float, float}]) is not. so, the current impl is not exactly right;
2) for zOS, i didn't see [2:F] is differentiated from [F,F] either;

I will suggest a correct implementation as follows in `parseStruct`:  under cases of digits (for array) and [{ (for nested struct/union), adding
```
if (isZos()) {
   *isAllDP = false;
   *isAllSP = false;
}
```
since you are aiming for ALL_SP/DP to represent complex(es) on zOS.

@r30shah or @joransiu  is my understanding of complex types on zOS correct?

Reply:
Hi @zl-wang , As @ChengJin01  as pointed in the comment, on z/OS ABI, a struct of size 8 bytes with two members , each is of same size (union of multiple floats only - The one that comment highlights), is considered complex. Similar for Double.

Reply:
thanks for the clarification. 
```
typedef struct { float a; float b}  Definitely_Complex
typedef union { Definitely_Complex a; float b;} Not_Complex_Anymore
typedef struct { Definitely_Complex a;} Not_Complex_Either
```
I guessed these are true according to @r30shah clarification, such that the implementation still needs to handle these and the array case correctly.

Reply:
on the other hand, 
`typedef struct { struct {float a;} b; float c; }  This_is_a_Complex
`
currently, this is handled correctly.